File: cpplogic/solver.cpp
#include "simulator.cpp"
#include <iostream>
#include <cassert>

class Solver {

private:
    Simulator Sim;

    state answer;
    steps answer_with_steps;

    piece object{1};

public:
    void AddPiece(piece &p) {
        Sim.AddPiece(p);
    };

    void ScanBlackPiece() {
        unit x, y;
        std::cin >> x >> y;

        piece p(1);
        p.loc = {x,y};
        object = p;

        AddPiece(p);
    };

    void ScanWhitePiece() {
        unit x, y;
        std::cin >> x >> y;

        piece p(1);
        p.loc = {x,y};

        AddPiece(p);
    };

    void PrintAnswer() {
        std::cout << "point: " << answer.point << std::endl;
        std::cout << "outOfBounds: " << answer.outOfBoudns << std::endl;
        std::cout << "vector: " << answer.p.vel.x <<' '<< answer.p.vel.y << std::endl;
    };

    void PrintAnswerStep() {
        Sim.UpdatePieceState(answer.p);
        Sim.RunWithStep(answer, answer_with_steps);
    };



    /*
    result에 결과를 담아 리턴하기
    매개변수나 리턴값 등 폭넓게 건드려도 됨
    속도를 단위벡터로 전달한 후 곱하는 상수값을 이분탐색
    실수 이분탐색을 잘 구현해야 하며 BS_CNT 만큼 반복
    */

    state MaxValNotCrossedOut(unit low, unit high, co unitVec) {
        /*
        밖으로 빠지지 않는 경우중 속도가 최대인 경우를 찾는 이분탐색 구현
        유효 오차가 특정 간격 이하가 될 때까지 반복
        */

        state result;

        while (high - low > BS_INTERVAL) {
            unit mid = (high + low) / 2;
            state candidate;
            piece p = object;
            p.vel = unitVec * mid;

            Sim.UpdatePieceState(p);
            Sim.Run(result);

            if (result.outOfBoudns == true) {
                high = mid;
            } else {
                low = mid;
                result = candidate;
            }
        }

        return result;
    };

    state MinVeMaxPoint(unit low, unit high, co unitVec, int point) {
        //밖으로 안빠지며 포인트가 최대인 경우중 속도가 최소를 찾는 이분탐색 구현
        state result;

        while (high - low > BS_INTERVAL) {
            unit mid = (high + low) / 2;
            state candidate;
            piece p = object;
            p.vel = unitVec * mid;

            Sim.UpdatePieceState(p);
            Sim.Run(result);

            assert(!result.outOfBoudns);
            assert(result.point <= point);

            if (result.point == point) {
                high = mid;
                result = candidate;
            } else {
                low = mid;
            }
        }

        return result;
    }

    void FindOptimal() {
        /*
        각도 브루트포스, 이분 탐색을 통해 최적을 찾아
        정답을 전역 변수 answer 에 저장
        */
        state result;

        for (int i=0; i<SEARCH_CNT; i++) {
            unit axis = (unit)i / SEARCH_CNT * 360;
            co unitVec = PHYSICS::UnitVector(axis);

            // 두 번의 binary search 를 통해 최적의 경우를 찾기
            result = MaxValNotCrossedOut(MIN_VEL, MAX_VEL, unitVec);
            unit max_vel = result.p.vel.Norm();
            result = MinVeMaxPoint(MIN_VEL, max_vel, unitVec, result.point);
        }

        answer = result;
    };
};

File: cpp_to_js/algos/solver.ts
import { BS_INTERVAL, Co, MAX_VELOCITY, MIN_VELOCITY, Piece, SEARCH_CNT, State } from "./api";
import { Simulator } from "./simulator";
import * as Physics from "./physics";

export class Solver {
    #sim: Simulator;
    #answer: State;
    #answerWithSteps: Array<Array<Piece>>;
    #object: Piece;

    constructor() {
        this.#sim = new Simulator();
        this.#answer = new State();
        this.#answerWithSteps = [];
        this.#object = new Piece(1);
    }

    addPiece(p: Piece) {
        this.#sim.addPiece(p.clone());
    }

    addBlackPiece(x: number, y: number) {
        const p = new Piece(1);
        p.loc.x = x;
        p.loc.y = y;
        this.#object = p;
        this.addPiece(p);
    }

    addWhitePiece(x: number, y: number) {
        const p = new Piece(0);
        p.loc.x = x;
        p.loc.y = y;
        this.addPiece(p);
    }
    
    printAnswer() {
        console.error(`point: ${this.#answer.point}`);
        console.error(`direction:\nx: ${this.#answer.p.vel.x}, y: ${this.#answer.p.vel.y}`);
        console.error(`outOfBounds: ${this.#answer.outOfBounds ? 'true' : 'false'}`);
    }

    printAnswerStep() {
        this.#sim.updatePieceState(this.#answer.p);
        this.#sim.runWithStep(this.#answer, this.#answerWithSteps);
    }

    findMaxValNotCrossedOut(low: number, high: number, unitVec: Co) {
        let result = new State();

        while (high - low > BS_INTERVAL) {
            const mid = (high + low) / 2;
            const candidate = new State();
            candidate.p = this.#object.clone();
            candidate.p.vel = unitVec.mulScalar(mid);

            this.#sim.updatePieceState(candidate.p);
            this.#sim.run(candidate);

            if (candidate.foul) {
                return candidate;
            }

            if (candidate.outOfBounds) {
                high = mid;
            } else {
                low = mid;
                if (result.point < candidate.point) {
                    result = candidate;
                }
            }
        }

        return result;
    }

    findOptimal() {
        for (let i = 0; i < SEARCH_CNT; ++i) {
            const axis = i * (360 / SEARCH_CNT);
            const unitVec = Physics.getUnitVector(axis);

            const result = this.findMaxValNotCrossedOut(MIN_VELOCITY, MAX_VELOCITY, unitVec);
            
            if (result.outOfBounds) {
                continue;
            }
            if (result.point > this.#answer.point) {
                this.#answer = result.clone();
            }
        }
    }
}